### ISSUE TITLE
[suggestion] object-shorthand to allow keys which aren't valid identifiers

### ISSUE THREAD

**Issue by spadgos** [eslint/eslint#3366]
There are some times when you need to assign a function to a key on an object, where the key is not a valid identifier (and hence, might need quoting)

```js
const x = {
  0: function () {}
  'x-y': function () {}
};
```

In these cases, to make sure that the `object-shorthand` method does not complain, you need to quote these, which leads to what might be considered some pretty ugly code:

```js
const x = {
  '0'() {},
  'x-y'() {}
}
```

I propose an additional option for that rule to allow the longhand style when the shorthand would require quoting.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nzakas** [comment]
Sounds fine to me. What would the option be called? Do you want to make the change?

**Comment by spadgos** [comment]
Hmm, maybe something in line with how the `quotes` rule has `"avoid-escape"`? This could be `"avoid-quotes"` ?

**Comment by nzakas** [comment]
I like it.

**Comment by alberto** [comment]
@spadgos are you interested in submitting a PR?

**Comment by lemonmade** [comment]
I'd be happy to take a stab at this. Is `avoid-quotes` an acceptable name for the option?

**Comment by ilyavolodin** [comment]
`avoid-quotes` sounds reasonable.

**Comment by lemonmade** [comment]
A few other questions: should this option only report keys that would be invalid identifiers? So, this reports:

```js
{'foo-bar'() {}}
```

but this does not:

```js
{'foo'() {}}
```

Seems like there is some overlap here with the `quote-props` rule.

I'm also wondering if there's an issue here in that computed properties aren't really checked one way or another. So, this passes even when set to `always`:

```js
{
  [foo]: function() {},
  ['x']: x,
}
```

Is this something I should fix here? In any case, we'd need to decide how this new option would deal with the following:

```js
{['foo']() {}}
```

**Comment by nzakas** [comment]
@lemonmade can you open a separate issue about computed properties? Let's not mudfy the waters with that discussion here. 

This option very simply says if the name of the function requires quotes in shorthand form, don't suggest it. So `"foo"` would not warn because it can be written without quotes but `"foo-bar"` cannot be written without quotes, so it should not warn.

**Comment by nzakas** [comment]
I commented on the PR, but I'm thinking `avoid-quotes` isn't the right option.

I think What we should do instead is add an `avoidQuotes` option on the options object, so:

```
object-shorthand: ["error", "always", { avoidQuotes: true}]
```

That way, you are still free to select " methods" or "properties" and still use this new option.

**Comment by lemonmade** [comment]
Thanks @nzakas. To clarify, do you still want the same behaviour of only warning for invalid identifiers, like `{'foo-bar': function() {}};`? I left a note about this on the PR, but the summary is: I think this additional option should ignore the object shorthand requirement for *any* stringified key (including `{'foo-bar': function () {}}`, `{'foo': function() {}}`, and even `{'x': x}`), and leave it to `quote-props` to handle which of those cases should be stringified keys in the first place.
